src/Features/LanguageServer/Protocol/Handler/RequestExecutionQueue.QueueItem.cs

```diff
@@ -85,40 +122,118 @@ internal partial class RequestExecutionQueue
                 TextDocument = textDocument;
-                HandleQueueFailure = handleQueueFailure;
-                CancellationToken = cancellationToken;
+                CombinedCancellationToken = combinedCancellationToken;
+            }
+
+            public static (IQueueItem, Task<TResponseType?>) Create(
+                bool mutatesSolutionState,
+                bool requiresLSPSolution,
+                ClientCapabilities clientCapabilities,
+                string? clientName,
+                string methodName,
+                TextDocumentIdentifier? textDocument,
+                TRequestType request,
+                IRequestHandler<TRequestType, TResponseType> handler,
+                Guid activityId,
+                ILspLogger logger,
+                RequestTelemetryLogger telemetryLogger,
+                CancellationToken combinedCancellationToken)
+            {
+                var queueItem = new QueueItem<TRequestType, TResponseType>(
+                    mutatesSolutionState,
+                    requiresLSPSolution,
+                    clientCapabilities,
+                    clientName,
+                    methodName,
+                    textDocument,
+                    request,
+                    handler,
+                    activityId,
+                    logger,
+                    telemetryLogger,
+                    combinedCancellationToken);
+
+                return (queueItem, queueItem._completionSource.Task);
             }
 
             /// <summary>
-            /// Processes the queued request. Exceptions that occur will be sent back to the requesting client, then re-thrown
+            /// Processes the queued request. Exceptions will be sent to the task completion source
+            /// representing the task that the client is waiting for, then re-thrown so that
+            /// the queue can correctly handle them depending on the type of request.
             /// </summary>
-            public async Task CallbackAsync(RequestContext? context, CancellationToken cancellationToken)
+            public async Task CallbackAsync(RequestContext? context)
             {
+                var cancellationToken = CombinedCancellationToken;
+
                 // Restore our activity id so that logging/tracking works.
                 Trace.CorrelationManager.ActivityId = ActivityId;
                 _logger.TraceStart($"{MethodName} - Roslyn");
                 try
                 {
-                    await _callbackAsync(context, cancellationToken).ConfigureAwait(false);
+                    // Check if cancellation was requested while this was waiting in the queue
+                    if (cancellationToken.IsCancellationRequested)
+                    {
+                        _logger.TraceInformation($"{MethodName} - Canceled while in queue");
+                        this.Metrics.RecordCancellation();
+
+                        _completionSource.SetCanceled();
```

Comment:
pass in the CT into SetCanceled?

Reply:
modified this bit to just `ThrowIfCancellationRequested` and let the catch below handle setting the result